### DISCUSSION TITLE
Как реализована выгрузка музыки из ВК?

### DISCUSSION THREAD

**Discussion by lgg** [Bizordec/vk-to-tgm#2] [Q&A]
Если я правильно понял, то текущее репо поддерживает выгрузку музыки? А можно этот компонент отдельно использовать вытянуть? @Bizordec, можешь, пожалуйста, подсказать - куда смотреть/копать в коде?

**Comment by Bizordec** [comment]
Выгрузка музыки реализована с помощью библиотеки [vkaudiotoken-python](https://github.com/vodka2/vkaudiotoken-python).
Запрос в API с токеном Kate Mobile возвращает ссылку на MP3, а с токеном официального VK клиента - на M3U8 (для скачивания требуется FFmpeg).

Конкретно в данном репозитории такая реализация:

1. Сначала берутся токены:
https://github.com/Bizordec/vk-to-tgm/blob/61f31756bdee1d0d41479ff834e7a35c59c284df/app/config.py#L221-L231
https://github.com/Bizordec/vk-to-tgm/blob/61f31756bdee1d0d41479ff834e7a35c59c284df/app/config.py#L37-L122

2. Затем создаются клиенты:
https://github.com/Bizordec/vk-to-tgm/blob/61f31756bdee1d0d41479ff834e7a35c59c284df/app/handlers/wall.py#L53-L70

3. Затем из API запрашиваются MP3 ссылки на аудио:
https://github.com/Bizordec/vk-to-tgm/blob/61f31756bdee1d0d41479ff834e7a35c59c284df/app/handlers/wall.py#L194-L202

4. И здесь уже идет скачивание аудио по 10 штук (максимум в Telegram на одно сообщение). В случае битой ссылки будет попытка скачивания по второму токену:
https://github.com/Bizordec/vk-to-tgm/blob/61f31756bdee1d0d41479ff834e7a35c59c284df/app/handlers/wall.py#L334-L344
https://github.com/Bizordec/vk-to-tgm/blob/61f31756bdee1d0d41479ff834e7a35c59c284df/app/utils/downloader.py#L74-L122

**Reply by lgg** [reply]
Спасибо большое за такой подробный ответ!
